### ISSUE TITLE
[CI] ConcurrentSeqNoVersioningIT.testSeqNoCASLinearizability fails

### ISSUE THREAD

**Issue by droberts195** [elastic/elasticsearch#42244]
`ConcurrentSeqNoVersioningIT.testSeqNoCASLinearizability` failed in https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=centos-6/428/console

The reported error is:

```
org.elasticsearch.versioning.ConcurrentSeqNoVersioningIT > testSeqNoCASLinearizability FAILED
    java.lang.RuntimeException: java.util.concurrent.TimeoutException
        at org.elasticsearch.versioning.ConcurrentSeqNoVersioningIT.testSeqNoCASLinearizability(ConcurrentSeqNoVersioningIT.java:190)

        Caused by:
        java.util.concurrent.TimeoutException
            at java.base/java.util.concurrent.CyclicBarrier.dowait(CyclicBarrier.java:258)
            at java.base/java.util.concurrent.CyclicBarrier.await(CyclicBarrier.java:436)
            at org.elasticsearch.versioning.ConcurrentSeqNoVersioningIT.testSeqNoCASLinearizability(ConcurrentSeqNoVersioningIT.java:174)

    java.lang.AssertionError: cluster failed to form after disruption was healed
```

I managed to reproduce this twice in a row on a CentOS 7 server using:

```
./gradlew :server:integTest --tests "org.elasticsearch.versioning.ConcurrentSeqNoVersioningIT.testSeqNoCASLinearizability" \
  -Dtests.seed=E8175DCFB75B80B2 \
  -Dtests.security.manager=true \
  -Dtests.locale=vun-TZ \
  -Dtests.timezone=Asia/Kashgar \
  -Dcompiler.java=12 \
  -Druntime.java=11
```

It seems to be seed-specific as this succeeded:

```
./gradlew :server:integTest --tests "org.elasticsearch.versioning.ConcurrentSeqNoVersioningIT.testSeqNoCASLinearizability"
```

**Comment by henningandersen** [comment]
This can be reproduced locally with provided sees, also in intellij. It seems a relocation hand-off cannot acquireAll permits on the shard. The test does stop indexing long before the test timeout happens (30s). Suspecting something does not release the permit. Will continue investigating.

**Comment by henningandersen** [comment]
The root cause seems to be the way `NetworkUnresponsive` works. It simply throws away any data send over affected connections, but never gives back any error. Most of ES depends on a transport action either succeeding or failing, but in this case, no response comes back for the operations, causing them to hang on the their shard permits until the test ends. But it never ends due to the relocation requiring all permits.

**Comment by henningandersen** [comment]
Snippets from logs:

```
[2019-05-27T12:16:50,389][INFO ][o.e.t.d.NetworkDisruption] [testSeqNoCASLinearizability] start disrupting (disruption type: network unresponsive, disrupted links: bridge partition (super connected node: [node_tm1], partition 1: [node_td4, node_tm2, node_td6, node_tm0] and partition 2: [node_td3, node_tc8, node_td5, node_td7]))

[2019-05-27T12:16:50,389][TRACE][o.e.a.b.TransportShardBulkAction] [node_td6] [[test][1]] sending op [indices:data/write/bulk[s]] to replica [test][1], node[qoXFAX30RPaCiP3zNG7dLQ (node_td3)], [R], s[STARTED], a[id=UC_z7ekyQ0S8PJojvg6vVQ] for request [BulkShardRequest [[test][1]] containing [index {[test][type][ID:5], source[{"value":627354534}]}]]
[2019-05-27T12:16:50,389][INFO ][o.e.v.ConcurrentSeqNoVersioningIT] [testSeqNoCASLinearizability] --> round 2
[2019-05-27T12:16:50,389][TRACE][o.e.a.b.TransportShardBulkAction] [node_td6] [[test][1]] sending op [indices:data/write/bulk[s]] to replica [test][1], node[fgbEqN63QFOBkECSnk2D2A (node_td7)], [R], s[STARTED], a[id=sSf7INhOShm2ue99FmaLug] for request [BulkShardRequest [[test][1]] containing [index {[test][type][ID:5], source[{"value":627354534}]}]]
[2019-05-27T12:16:50,389][TRACE][o.e.a.b.TransportShardBulkAction] [node_td6] [[test][1]] sending op [indices:data/write/bulk[s]] to replica [test][1], node[I2RAZUotR_WX8bHsj1KkUw (node_td5)], [R], s[STARTED], a[id=yiunNSBLRmmIFT_9gfWXRw] for request [BulkShardRequest [[test][1]] containing [index {[test][type][ID:5], source[{"value":627354534}]}]]

[2019-05-27T12:16:56,074][INFO ][o.e.t.d.NetworkDisruption] [testSeqNoCASLinearizability] start disrupting (disruption type: network disconnects, disrupted links: two partitions (partition 1: [node_td4, node_tc8, node_tm0] and partition 2: [node_td3, node_tm2, node_td6, node_td5, node_tm1, node_td7]))

[2019-05-27T12:17:00,076][INFO ][o.e.t.InternalTestCluster] [testSeqNoCASLinearizability] Clearing active scheme network disruption (disruption type: network disconnects, disrupted links: two partitions (partition 1: [node_td4, node_tc8, node_tm0] and partition 2: [node_td3, node_tm2, node_td6, node_td5, node_tm1, node_td7])), expected healing time 0s
[2019-05-27T12:17:00,076][INFO ][o.e.t.d.NetworkDisruption] [testSeqNoCASLinearizability] stop disrupting (disruption scheme: network disconnects, disrupted links: two partitions (partition 1: [node_td4, node_tc8, node_tm0] and partition 2: [node_td3, node_tm2, node_td6, node_td5, node_tm1, node_td7]))

[2019-05-27T12:17:00,386][TRACE][o.e.i.r.RecoverySourceHandler] [node_td6] [test][1][recover to node_td4] performing relocation hand-off
```
High-level summary:
1. Connections are already established between all relevant nodes 
2. An unresponsive partition where td6 cannot send to its replicas td3, td5, td7 is added.
3. Replica operations are send from td6 to td3, td5, td7.
4. A few seconds later network disruptions are cleared.
5. A network disconnects disruption is added, td4 is isolated away from master and all data nodes. td6 can communicate to its 3 replicas.
6. td4 is assumed dead by master and its shards are unassigned.
7. A few seconds later, network disruptions are cleared.
8. td4 is considered live again (reconnects).
9. Rebalancing puts shards on td4, ending in a hand in relocation hand-off from td6 to td4 due to never seeing any response from the transport requests in step 3.

**Comment by ywelsch** [comment]
> The root cause seems to be the way NetworkUnresponsive works. It simply throws away any data send over affected connections, but never gives back any error.

Can we change `NetworkUnresponsive` so that it does not simply throw data away, but caches it and processes it later when the disruption is healed (similar as in `public void addUnresponsiveRule(TransportAddress transportAddress, final TimeValue duration) {`) /cc: @DaveCTurner

**Comment by henningandersen** [comment]
Another option is to simply close it afterwards, to make it considerably different from the delay disruption. Otherwise, these will behave identically as long as the delay is longer than the disrupt period?

**Comment by DaveCTurner** [comment]
This specific situation relates to #41307 but I do agree that the `NetworkUnresponsive` rule is not a realistic disruption as implemented today. I think we can do more to make these disruptions cover more cases, but closing all the connections at the end of the partition already seems like an improvement.

**Comment by henningandersen** [comment]
A quick summary of the 3 network link disruption types we have today:

* NetworkDisconnect: any connection touched during disruption will be closed and any connection attempt will fail.
* NetworkDelay: delays any send request attempts in a period, sends them once disruption is cleared. Likewise, connection attempts are delayed in the same period.
* NetworkUnresponsive: currently swallows all send requests. Connection attempts are rejected immediately.

**Comment by dimitris-athanasiou** [comment]
We got another one of these but this time it OOM'd.

Build link: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.2+multijob-unix-compatibility/os=opensuse/39/consoleText

Reproduce:

```
./gradlew :server:integTest --tests "org.elasticsearch.versioning.ConcurrentSeqNoVersioningIT.testSeqNoCASLinearizability" -Dtests.seed=ABFE44FD823A8E69 -Dtests.security.manager=true -Dtests.locale=zh-HK -Dtests.timezone=Pacific/Honolulu -Dcompiler.java=12 -Druntime.java=8
```

Failure:
```
org.elasticsearch.versioning.ConcurrentSeqNoVersioningIT > testSeqNoCASLinearizability FAILED
    java.lang.OutOfMemoryError: GC overhead limit exceeded
        at __randomizedtesting.SeedInfo.seed([ABFE44FD823A8E69:70306A979828ACDD]:0)
        at org.apache.lucene.util.FixedBitSet.clone(FixedBitSet.java:480)
        at org.elasticsearch.cluster.coordination.LinearizabilityChecker.isLinearizable(LinearizabilityChecker.java:253)
        at org.elasticsearch.cluster.coordination.LinearizabilityChecker.lambda$isLinearizable$0(LinearizabilityChecker.java:233)
        at org.elasticsearch.cluster.coordination.LinearizabilityChecker$$Lambda$3170/2121402563.test(Unknown Source)
        at java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
        at java.util.Collections$2.tryAdvance(Collections.java:4717)
        at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
        at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
        at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.allMatch(ReferencePipeline.java:454)
        at org.elasticsearch.cluster.coordination.LinearizabilityChecker.isLinearizable(LinearizabilityChecker.java:233)
        at org.elasticsearch.versioning.ConcurrentSeqNoVersioningIT$Partition.isLinearizable(ConcurrentSeqNoVersioningIT.java:450)
        at org.elasticsearch.versioning.ConcurrentSeqNoVersioningIT$Partition.assertLinearizable(ConcurrentSeqNoVersioningIT.java:462)
        at org.elasticsearch.versioning.ConcurrentSeqNoVersioningIT$$Lambda$3166/1664813076.accept(Unknown Source)
        at java.util.ArrayList.forEach(ArrayList.java:1257)
        at org.elasticsearch.versioning.ConcurrentSeqNoVersioningIT.testSeqNoCASLinearizability(ConcurrentSeqNoVersioningIT.java:199)
```

**Comment by henningandersen** [comment]
Build scan for last reported failure is here: https://scans.gradle.com/s/elpabmodig2jm

**Comment by henningandersen** [comment]
This second CI failure can be reproduced with the seed from above on both 7.2 and master. The problem seems to be long and complex histories (many timeouts) making the linearizability checker choke on it. We could improve the caching strategy for large histories, but this specific problem does not appear with the unresponsive network simulation fix from #42579 so will leave this for now.
